membership-attribute-service/app/services/PaymentFailureAlerter.scala

```diff
@@ -0,0 +1,85 @@
+package services
+
+import java.util.Locale
+
+import com.gu.memsub.subsv2.Subscription
+import com.gu.memsub.subsv2.SubscriptionPlan.AnyPlan
+import com.gu.zuora.api.{RegionalStripeGateways, StripeAUMembershipGateway, StripeUKMembershipGateway}
+import com.gu.zuora.rest.ZuoraRestService.{AccountObject, AccountSummary, PaymentMethodId, PaymentMethodResponse}
+import org.joda.time.DateTime
+import org.joda.time.format.DateTimeFormat
+
+import scala.concurrent.{ExecutionContext, Future}
+import scalaz.syntax.std.boolean._
+import scalaz.{Disjunction, \/}
+
+
+object PaymentFailureAlerter {
+
+  private def accountObject(accountSummary: AccountSummary) =
+    AccountObject(
+      Id = accountSummary.id,
+      Balance = accountSummary.balance,
+      Currency = accountSummary.currency,
+      DefaultPaymentMethodId = accountSummary.defaultPaymentMethod.map(_.id),
+      PaymentGateway = accountSummary.billToContact.country.map(RegionalStripeGateways.getGatewayForCountry)
+    )
+
+
+  def membershipAlertText(accountSummary: AccountSummary, subscription: Subscription[AnyPlan],
+                          paymentMethodGetter: PaymentMethodId => Future[String \/ PaymentMethodResponse])(implicit ec: ExecutionContext) : Future[Option[String]] = {
+
+    def expectedAlertText: Future[Option[String]] = {
+      val formatter = DateTimeFormat.forPattern("d MMMM yyyy").withLocale(Locale.ENGLISH)
+      val paymentMethodLatestDateFormatted: Future[String \/ String] = accountSummary.defaultPaymentMethod.map(_.id) match {
+        case Some(id) => {
+          val paymentMethod: Future[Disjunction[String, PaymentMethodResponse]] = paymentMethodGetter(id) fallbackTo Future.successful(\/.left("Failed to get payment method"))
+          paymentMethod map { paymentMethodDisjunction: Disjunction[String, PaymentMethodResponse] => paymentMethodDisjunction map { pm => pm.lastTransactionDateTime.toString(formatter)}}
```

Comment:
if you just want to map through both layers, consider wrap in EitherT, then map, then do .run to get it back again.  May be neater?

Reply:
I'd rather leave the formatting to a string until the last moment, keep it as a proper type. This avoids having a String \/ String type too, which reduces the chance of error.
so just remove the `.toString(formatter)` from here, fix up the types, and do it on line 44/45 instead